### ISSUE TITLE
ESLint doesn't exit on failOnError();

### ISSUE THREAD

**Issue by ahmedelgabri** [adametry/gulp-eslint#59]
So I'm creating gulp tasks dynamically like this
```js
var gulp           = require('gulp');
var eslint         = require('gulp-eslint');
var apps           = require('../config').browserify;
var createPaths    = require('../helpers/create-paths');
var lintTasks      = [];
var watchLintTasks = [];

var lint = function( project ) {
    return function() {
        var srcFiles = createPaths.watchPath(project, 'js');

        return gulp.src(srcFiles)
            .pipe(eslint())
            .pipe(eslint.format())
            .pipe(eslint.failOnError());
    };
};

apps.map(function(app){
    lintTasks.push('lint:' + app);
    watchLintTasks.push('watch:lint:' + app);

    gulp.task('lint:' + app, lint(app));
    gulp.task('watch:lint:' + app, function(){
        gulp.watch(createPaths.watchPath(app, 'js'), ['lint:'+ app]);
    });
});

gulp.task('lintScripts', lintTasks);
gulp.task('watch:lintScripts', watchLintTasks);
```

And then in my build task I'm running `gulp.task('build', ['lintScripts', ...]);`

When I run `gulp build` it doesn't exit on errors I tried many solutions from #12 but still not exiting when it fails any idea how to fix this?

**Comment by juhasz** [comment]
It could be the same problem as #36.

**Comment by adametry** [comment]
It looks like you're running your lint and build tasks [in parallel](https://github.com/gulpjs/gulp/blob/master/docs/recipes/running-tasks-in-series.md). If you want tasks to wait for the `lintScripts` task to pass (and not run if it fails), make sure those tasks are dependent on `lintScripts` (gulp will keep.

Here's an example of tasks running in (dependency) sequence:
```javascript
// ...after the lintTasks are prepared
gulp.task('lintScripts', lintTasks, function () {
     console.log('All lint tasks succeeded');
});

gulp.task('postLintTask', ['lintScripts'], function () {
    console.log('running post-link task now...');
});

gulp.task('build', ['lintScripts','postLintTask'], function () {
    console.log('running build task now...');
});

```

Otherwise, your code works fine for me (with a bit of a guess at what `createPaths.watchPath` does); though, because all the linting happens in parallel, the process may get more than one error before exiting with a code of 1. One way to control that is to merge the the streams between `eslint()` and `eslint.format()` (can provide detail, if interested).

**Comment by ahmedelgabri** [comment]
@adametry Thanks for the explanation, it makes sense. It'll be really nice if you provide more details about merging streams.

**Comment by adametry** [comment]
Merging streams works by piping data from multiple streams through one point. The `merge-stream` NPM package is a strait forward way to do it. Gulp has a simple [demonstration] (https://github.com/gulpjs/gulp/blob/master/docs/recipes/using-multiple-sources-in-one-task.md).

Merging streams works well when the various input streams start differently (e.g., browserify, *-to-JS transpilers, varying sources) yet end the same (e.g., gulp.dest, reporting). Applied to the original example, it might look like:

```javascript
var gulp = require('gulp');
var eslint = require('gulp-eslint');
var merge = require('merge-stream');
var apps = require('../config').browserify;
var createPaths = require('../helpers/create-paths');

// This creates a stream of linted files (createPaths + gulp.src + eslint)
var createLintedSrcStream = function(app) {
	var paths = createPaths.watchPath(app, 'js');
	return gulp.src(paths)
		.pipe(eslint());
}

// create a lint task for each app
apps.forEach(function(app) {

	gulp.task('lint:' + app, function lintAndReportTask() {

		// create a (linted) src stream. Then, print the result, fail if any file had an error
		return createLintedSrcStream(app)
			.pipe(eslint.format())
			.pipe(eslint.failAfterError());
	});
});

// lint all apps
gulp.task('lintScripts', function() {

	// this will start a separate (linted) src streams for each app
	var allStreams = apps.map(createLintedSrcStream);

	// The files from each stream will be merged into a single stream
	return merge(allStreams)
		//	Now it can be treated like any other gulp src stream
		.pipe(eslint.format())
		.pipe(eslint.failAfterError());

});
```

Admittedly, after writing this example, it does seem needlessly complex in this case. Your original version may very well be perfect. If the tasks for linting each app's srcFiles are the same, a slight variation may work as well:

```javascript
var gulp = require('gulp');
var eslint = require('gulp-eslint');
var merge = require('merge-stream');
var apps = require('../config').browserify;
var createPaths = require('../helpers/create-paths');

// This is a simple alternative to the original example
var createLinterAndReporterTask = function(appPaths) {
	// return a function that will lint a pre-defined set of paths/globs
	return function() {
		return gulp.src(appPaths)
			.pipe(eslint())
			.pipe(eslint.format())
			.pipe(eslint.failAfterError());
	};
};

var allAppPaths = apps.reduce(function(allPaths, app) {
	// get app src path/glob
	var appPaths = createPaths.watchPath(app, 'js');

	// app-specific task
	gulp.task('lint-paths:' + app, createLinterAndReporterTask(appPaths));

	// concat a list of all app files
	return allPaths.concat(appPaths);

}, []);

// create a task using all app paths
gulp.task('lint-all-paths', createLinterAndReporterTask(allAppPaths));
```

Any way you go with it, you should get an exit error code if any lint reports include an error.

Also, in these examples I use `format` with `failAfterError` since these examples don't include `gulp.watch` (which creates a never-ending stream). Alternatively, `formatEach` with `failOnError` work well with gulp.watch, since they act upon each file (rather than waiting for the end of the stream to report/fail). Other combinations of these methods can be a bit awkward.

**Comment by ahmedelgabri** [comment]
Thanks a lot @adametry!

**Comment by MatthewHerbst** [comment]
I'm having this same issue when running the lint task from a watch task.

In my gulpfile.js
```javascript
function lintServer() {
  return gulp.src([paths.app.server])
      .pipe(eslint())
      .pipe(eslint.format())
      .pipe(eslint.failAfterError());
}

function watchServer() {
  gulp.watch([paths.app.server], ['lint:server']);
}
```

The command lint reruns, but doesn't even report, let alone exit on, errors.
